File: bridgeAddon/addon/globalPlugins/nvdaMcpBridge/domain/controllers/commands/hello.py
# nvdaMcpBridge domain -- HelloHandler: the bootstrap command.
# Copyright (C) 2026 Marlon Brandao de Sousa. GPL-2. See COPYING.txt.
#
# ROLE: command handler for `hello` -- the one command valid before the handshake
# completes, and the one that BUILDS the session. Unlike the operational handlers
# (which only read a ready SessionContext), hello is wired with the AdapterFactory
# and the NVDA version, and it populates the context: builds the mode-specific
# adapters, creates the buffers, and starts capture. There is NO synth swap: the
# reader's real synth stays loaded in every mode; silent mode suppresses NVDA's
# speak() output at the speech source instead (see nvda_silent_speech_source), so
# NVDA and other add-ons keep seeing their configured synth as valid and active.
#
# On a protocol-version mismatch it raises CommandError before touching anything
# -- the factory is never called -- and the Session ends the handshake.

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from .... import protocol
from ...entities.braille_buffer import BrailleBuffer
from ...entities.speech_buffer import SpeechBuffer
from .command_handler import CommandError, CommandHandler

if TYPE_CHECKING:
	from ....protocol import Capability, ReaderInfo
	from ...ports.adapter_factory import AdapterFactory
	from .session_context import SessionContext


class HelloHandler(CommandHandler):
	available_before_hello = True

	def __init__(
		self,
		factory: AdapterFactory,
		reader: ReaderInfo,
		capabilities: list[Capability],
	) -> None:
		self._factory = factory
		self._reader = reader
		self._capabilities = capabilities

	def execute(self, ctx: SessionContext, request: protocol.Request) -> Any:
		params = protocol.from_dict(protocol.HelloParams, request.params)
		if params.protocolVersion != protocol.PROTOCOL_VERSION:
			raise CommandError(
				f"protocol version mismatch: bridge speaks {protocol.PROTOCOL_VERSION}, "
				f"client sent {params.protocolVersion}"
			)
		ctx.transcript.open()
		# Capture is always on (spec 0009); logLevel, if set, additionally bumps
		# NVDA's own verbosity for the session -- restored at teardown.
		ctx.log_capture.start(params.logLevel)
		adapters = self._factory.build(params.mode)
		# Installed before starting capture, so teardown can stop the sources even
		# if a start() below raises.
		ctx.adapters = adapters

		# No exact-finish signal: silent mode suppresses at the speak() filter, so
		# there is no synth "done speaking" to key off; both modes use the buffer's
		# elapsed-time heuristic.
		speech = SpeechBuffer(ctx.clock, exact_finish=False)
		braille = BrailleBuffer(ctx.clock)
		speech.set_observer(ctx.transcript.speech)
		ctx.speech = speech
		ctx.braille = braille
		adapters.speech_source.start(speech)
		adapters.braille_source.start(braille)

		# The reader's real synth stays loaded in every mode; just report it.
		synth = ctx.announcer.current_synth()
		ctx.transcript.session_opened(params.mode, synth)

		return protocol.HelloResult(
			protocolVersion=protocol.PROTOCOL_VERSION,
			reader=self._reader,
			capabilities=self._capabilities,
			mode=params.mode,
			synth=synth,
			logPath=ctx.transcript.path,
			nvdaLogPath=ctx.log_capture.path,
		)

File: bridges/voiceover/Sources/VoiceOverBridgeDomain/Controllers/Commands/Hello.swift
// ROLE: controller for `hello`, the only command legal before the handshake completes.
// HOLDS: the AdapterFactory, the reader identity, the served capabilities and the bridge version.
// BUILT BY: Registry.
// POPULATES: the SessionContext's mode, persona and adapter set.

import ScreenReaderWire

public final class HelloHandler: CommandHandler {
	public let availableBeforeHello = true

	private let factory: any AdapterFactory
	private let reader: ReaderInfo
	private let capabilities: [Capability]
	private let bridgeVersion: String

	public init(
		factory: any AdapterFactory,
		reader: ReaderInfo,
		capabilities: [Capability],
		bridgeVersion: String
	) {
		self.factory = factory
		self.reader = reader
		self.capabilities = capabilities
		self.bridgeVersion = bridgeVersion
	}

	public func execute(_ context: SessionContext, _ request: Request) throws -> any Encodable {
		let params = try request.params(as: HelloParams.self)
		guard ProtocolVersion.supports(params.protocolVersion) else {
			throw CommandError(
				"protocol version mismatch: bridge speaks \(ProtocolVersion.current), "
					+ "client sent \(params.protocolVersion)"
			)
		}

		// Recorded before anything else can fail, so any evidence the session produces is attributable.
		context.persona = params.persona
		context.mode = params.mode
		context.transcript.open()

		let adapters: AdapterSet
		do {
			adapters = try factory.build(mode: params.mode)
		} catch let refusal as AdapterFactoryError {
			throw CommandError(refusal.description)
		}
		// Installed before capture starts, so teardown can stop what was started if a later step throws.
		context.adapters = adapters

		// One buffer per session: indices that outlived a session would mislead the next agent.
		let speech = SpeechBuffer(clock: context.clock)
		// Wired before anything is captured, so a crashed run still leaves a record of what the reader said.
		speech.setObserver { [transcript = context.transcript] text in transcript.speech(text) }
		context.speech = speech
		adapters.speechSource.start(speech)

		// After capture is listening: pointing the reader at the capture voice starts utterances, and the
		// setup's last rung requires one to reach this buffer.
		try ReaderEdgeSetup(adapters: adapters, context: context, speech: speech).establish()

		context.transcript.sessionOpened(
			mode: params.mode.rawValue,
			voice: captureVoiceName,
			persona: params.persona
		)

		return HelloResult(
			protocolVersion: ProtocolVersion.current,
			reader: reader,
			capabilities: capabilities,
			mode: params.mode,
			synth: selectedSynth(adapters.providerLifecycle),
			logPath: context.transcript.logPath,
			bridgeVersion: bridgeVersion,
			// Sent only because this build announces the `guidance` capability; without it the field must be
			// omitted.
			guidance: try GetGuidanceHandler.guidance(for: params.persona),
			// Sent in both modes, since it describes the machine; never settable over the wire.
			silenceCap: SilenceCapInfo(
				enabled: context.silenceCapPolicy.enabled,
				warnAfterSeconds: context.silenceCapPolicy.warnAfter,
				liftAfterSeconds: context.silenceCapPolicy.liftAfter
			),
			attended: context.attended
		)
	}

	private func selectedSynth(_ lifecycle: any ProviderLifecycle) -> String {
		guard let selected = lifecycle.selectedVoice() else { return captureVoiceName }
		return selected.isCaptureVoice ? captureVoiceName : selected.identifier
	}
}

let captureVoiceName = "screen-readers-mcp capture voice"
